kube-controller-manager panic happend when add StatefulSetAutoDeletePVC feature gate and create statefuleset with volumeClaimTemplates

### What happened?


The replicas is 3 but just create 1 pod 'web-0' and one pvc, then kube-controller-manager panic happend.
```
I1025 09:12:25.263508       1 shared_informer.go:262] Caches are synced for crt configmap
I1025 09:12:25.263537       1 shared_informer.go:262] Caches are synced for stateful set
I1025 09:12:25.265883       1 shared_informer.go:262] Caches are synced for certificate-csrsigning-kubelet-client
I1025 09:12:25.265911       1 shared_informer.go:262] Caches are synced for certificate-csrsigning-kubelet-serving
I1025 09:12:25.266001       1 shared_informer.go:262] Caches are synced for endpoint
I1025 09:12:25.266799       1 shared_informer.go:262] Caches are synced for certificate-csrsigning-kube-apiserver-client
I1025 09:12:25.268054       1 shared_informer.go:262] Caches are synced for certificate-csrsigning-legacy-unknown
I1025 09:12:25.268085       1 shared_informer.go:262] Caches are synced for TTL
E1025 09:12:25.269192       1 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 1947 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x3d40ac0?, 0x6fcaeb0})
	vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc001aeb970?})
	vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49 +0x75
panic({0x3d40ac0, 0x6fcaeb0})
	/usr/local/go/src/runtime/panic.go:884 +0x212
k8s.io/kubernetes/pkg/controller/statefulset.claimOwnerMatchesSetAndPod(0x0?, 0xc0027b8000, 0xc00269f658?)
	pkg/controller/statefulset/stateful_set_utils.go:159 +0x42
k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulPodControl).ClaimsMatchRetentionPolicy(0xc000566220, 0xc0027b8000, 0xc001f08e20)
	pkg/controller/statefulset/stateful_pod_control.go:222 +0x1e9
k8s.io/kubernetes/pkg/controller/statefulset.(*defaultStatefulSetControl).updateStatefulSet(0xc000569a80, {0x4dcd730, 0xc0002374c0}, 0xc001e7a000, 0xc001fa2750, 0xc001fa2750, 0x0, {0xc0000125a0, 0x1, 0x1})
	pkg/controller/statefulset/stateful_set_control.go:470 +0x136b
k8s.io/kubernetes/pkg/controller/statefulset.(*defaultStatefulSetControl).performUpdate(0x15ed520?, {0x4dcd730, 0xc0002374c0}, 0xc001170500?, {0xc0000125a0, 0x1, 0x1}, {0xc0000125c0, 0x1, 0x1})
	pkg/controller/statefulset/stateful_set_control.go:107 +0xe7
k8s.io/kubernetes/pkg/controller/statefulset.(*defaultStatefulSetControl).UpdateStatefulSet(0xc0008f1474?, {0x4dcd730, 0xc0002374c0}, 0xc0003e5a28?, {0xc0000125a0, 0x1, 0x1})
	pkg/controller/statefulset/stateful_set_control.go:88 +0x138
k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulSetController).syncStatefulSet(0xc0009761b0, {0x4dcd730, 0xc0002374c0}, 0xc0003e5a28, {0xc0000125a0, 0x1, 0x1})
	pkg/controller/statefulset/stateful_set.go:481 +0x18a
k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulSetController).sync(0xc0009761b0, {0x4dcd730, 0xc0002374c0}, {0xc00021c580, 0xb})
	pkg/controller/statefulset/stateful_set.go:473 +0x425
k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulSetController).processNextWorkItem(0xc0009761b0, {0x4dcd730, 0xc0002374c0})
	pkg/controller/statefulset/stateful_set.go:421 +0x11b
k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulSetController).worker(0xc001e38240?, {0x4dcd730, 0xc0002374c0})
	pkg/controller/statefulset/stateful_set.go:432 +0x39
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:190 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x2273616369?)
	vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:157 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x73752f151a001077?, {0x4da2be0, 0xc001e38240}, 0x1, 0xc000b2bf80)
	vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:158 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc320a2073796177?, 0x3b9aca00, 0x0, 0x52?, 0x18a0001820072?)
	vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:135 +0x89
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x4dcd730, 0xc0002374c0}, 0xc0024f0b40, 0x73656c69460a3263?, 0xa091a6d65747379?, 0x7?)
	vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:190 +0x99
k8s.io/apimachinery/pkg/util/wait.UntilWithContext({0x4dcd730?, 0xc0002374c0?}, 0x1b2cde6?, 0x2e1a0a480a400008?)
	vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:101 +0x2b
created by k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulSetController).Run
	pkg/controller/statefulset/stateful_set.go:160 +0x353
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x163b2a2]

goroutine 1947 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc001aeb970?})
	vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:56 +0xd7
panic({0x3d40ac0, 0x6fcaeb0})
	/usr/local/go/src/runtime/panic.go:884 +0x212
k8s.io/kubernetes/pkg/controller/statefulset.claimOwnerMatchesSetAndPod(0x0?, 0xc0027b8000, 0xc00269f658?)
	pkg/controller/statefulset/stateful_set_utils.go:159 +0x42
k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulPodControl).ClaimsMatchRetentionPolicy(0xc000566220, 0xc0027b8000, 0xc001f08e20)
	pkg/controller/statefulset/stateful_pod_control.go:222 +0x1e9
k8s.io/kubernetes/pkg/controller/statefulset.(*defaultStatefulSetControl).updateStatefulSet(0xc000569a80, {0x4dcd730, 0xc0002374c0}, 0xc001e7a000, 0xc001fa2750, 0xc001fa2750, 0x0, {0xc0000125a0, 0x1, 0x1})
	pkg/controller/statefulset/stateful_set_control.go:470 +0x136b
k8s.io/kubernetes/pkg/controller/statefulset.(*defaultStatefulSetControl).performUpdate(0x15ed520?, {0x4dcd730, 0xc0002374c0}, 0xc001170500?, {0xc0000125a0, 0x1, 0x1}, {0xc0000125c0, 0x1, 0x1})
	pkg/controller/statefulset/stateful_set_control.go:107 +0xe7
k8s.io/kubernetes/pkg/controller/statefulset.(*defaultStatefulSetControl).UpdateStatefulSet(0xc0008f1474?, {0x4dcd730, 0xc0002374c0}, 0xc0003e5a28?, {0xc0000125a0, 0x1, 0x1})
	pkg/controller/statefulset/stateful_set_control.go:88 +0x138
k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulSetController).syncStatefulSet(0xc0009761b0, {0x4dcd730, 0xc0002374c0}, 0xc0003e5a28, {0xc0000125a0, 0x1, 0x1})
	pkg/controller/statefulset/stateful_set.go:481 +0x18a
k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulSetController).sync(0xc0009761b0, {0x4dcd730, 0xc0002374c0}, {0xc00021c580, 0xb})
	pkg/controller/statefulset/stateful_set.go:473 +0x425
k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulSetController).processNextWorkItem(0xc0009761b0, {0x4dcd730, 0xc0002374c0})
	pkg/controller/statefulset/stateful_set.go:421 +0x11b
k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulSetController).worker(0xc001e38240?, {0x4dcd730, 0xc0002374c0})
	pkg/controller/statefulset/stateful_set.go:432 +0x39
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:190 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x2273616369?)
	vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:157 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x73752f151a001077?, {0x4da2be0, 0xc001e38240}, 0x1, 0xc000b2bf80)
	vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:158 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc320a2073796177?, 0x3b9aca00, 0x0, 0x52?, 0x18a0001820072?)
	vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:135 +0x89
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x4dcd730, 0xc0002374c0}, 0xc0024f0b40, 0x73656c69460a3263?, 0xa091a6d65747379?, 0x7?)
	vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:190 +0x99
k8s.io/apimachinery/pkg/util/wait.UntilWithContext({0x4dcd730?, 0xc0002374c0?}, 0x1b2cde6?, 0x2e1a0a480a400008?)
	vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:101 +0x2b
created by k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulSetController).Run
	pkg/controller/statefulset/stateful_set.go:160 +0x353
```

### What did you expect to happen?

Should create 3 pod and 3 pvc

### How can we reproduce it (as minimally and precisely as possible)?

add '--feature-gates=StatefulSetAutoDeletePVC=true' arg for kube-controller-manager, then create statefulset with 2 replicas and volumeClaimTemplates.

```
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: web
spec:
  selector:
    matchLabels:
      app: nginx 
  serviceName: "nginx"
  replicas: 3 
  minReadySeconds: 10 
  template:
    metadata:
      labels:
        app: nginx 
    spec:
      terminationGracePeriodSeconds: 10
      containers:
      - name: nginx
        image: nginx:latest
        ports:
        - containerPort: 80
          name: web
        volumeMounts:
        - name: www
          mountPath: /usr/share/nginx/html
  volumeClaimTemplates:
  - metadata:
      name: www
    spec:
      accessModes: [ "ReadWriteOnce" ]
      storageClassName: "csi-rbd-sc"
      resources:
        requests:
          storage: 1Gi
```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.3", GitCommit:"434bfd82814af038ad94d62ebe59b133fcb50506", GitTreeState:"clean", BuildDate:"2022-10-12T10:57:26Z", GoVersion:"go1.19.2", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.3", GitCommit:"434bfd82814af038ad94d62ebe59b133fcb50506", GitTreeState:"clean", BuildDate:"2022-10-12T10:49:09Z", GoVersion:"go1.19.2", Compiler:"gc", Platform:"linux/amd64"}
```

</details>


### Cloud provider

<details>

</details>


### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
$ uname -a
# Linux dps-master 5.4.0-100-generic #113-Ubuntu SMP Thu Feb 3 18:43:29 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>


### Install tools

<details>

</details>


### Container runtime (CRI) and version (if applicable)

<details>

</details>


### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

--- a/pkg/controller/statefulset/stateful_set_control_test.go
+++ b/pkg/controller/statefulset/stateful_set_control_test.go
@@ -123,6 +123,8 @@ func runTestOverPVCRetentionPolicies(t *testing.T, testName string, testFn func(
 			WhenScaled:  apps.DeletePersistentVolumeClaimRetentionPolicyType,
 			WhenDeleted: apps.DeletePersistentVolumeClaimRetentionPolicyType,
 		},
+		// tests the case when no policy is set.
+		nil,
 	} {
 		subtestName := pvcDeletePolicyString(policy) + "/StatefulSetAutoDeletePVCEnabled"
 		if testName != "" {
@@ -136,6 +138,9 @@ func runTestOverPVCRetentionPolicies(t *testing.T, testName string, testFn func(
 }
 
 func pvcDeletePolicyString(policy *apps.StatefulSetPersistentVolumeClaimRetentionPolicy) string {
+	if policy == nil {
+		return "nullPolicy"
+	}
 	const retain = apps.RetainPersistentVolumeClaimRetentionPolicyType
 	const delete = apps.DeletePersistentVolumeClaimRetentionPolicyType
 	switch {

--- a/pkg/controller/statefulset/stateful_set_utils.go
+++ b/pkg/controller/statefulset/stateful_set_utils.go
@@ -149,7 +149,7 @@ func getPersistentVolumeClaimRetentionPolicy(set *apps.StatefulSet) apps.Statefu
 // claimOwnerMatchesSetAndPod returns false if the ownerRefs of the claim are not set consistently with the
 // PVC deletion policy for the StatefulSet.
 func claimOwnerMatchesSetAndPod(claim *v1.PersistentVolumeClaim, set *apps.StatefulSet, pod *v1.Pod) bool {
-	policy := set.Spec.PersistentVolumeClaimRetentionPolicy
+	policy := getPersistentVolumeClaimRetentionPolicy(set)
 	const retain = apps.RetainPersistentVolumeClaimRetentionPolicyType
 	const delete = apps.DeletePersistentVolumeClaimRetentionPolicyType
 	switch {
@@ -211,7 +211,7 @@ func updateClaimOwnerRefForSetAndPod(claim *v1.PersistentVolumeClaim, set *apps.
 	updateMeta(&podMeta, "Pod")
 	setMeta := set.TypeMeta
 	updateMeta(&setMeta, "StatefulSet")
-	policy := set.Spec.PersistentVolumeClaimRetentionPolicy
+	policy := getPersistentVolumeClaimRetentionPolicy(set)
 	const retain = apps.RetainPersistentVolumeClaimRetentionPolicyType
 	const delete = apps.DeletePersistentVolumeClaimRetentionPolicyType
 	switch {

